### PR TITLE
Implementing IRQAllocator in vmm

### PR DESCRIPTION
**Pull request by ramyak-mehra**
We needed IRQAllocator for adding support
for multiple devices. IRQ_NUM for each platform is now passed
down through VMConfig. default value are used.

Signed-off-by: Ramyak mehra <rmehra_be19@thapar.edu>

### Summary of the PR

*Please summarize here why the changes in this PR are needed.*

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] Any newly added `unsafe` code is properly documented.

### CODE REVIEW

**Comment by ramyak-mehra** [comment]
This PR is in continuation to a previous PR #227, whose branch was accidentally deleted thus resulting in PR automatically being closed.

**Comment by andreeaflorescu** [comment]
@ramyak-mehra can you please add a separate commit to fix the `cargo audit` warning? The fix is pretty simple, you just need to update the version of `clap` to 3.2.17, which is the latest release. In the commit you need to have both the `Cargo.toml` update as well as the `Cargo.lock` update.

**Comment by andreeaflorescu** [comment]
This is where the version update is needed: https://github.com/rust-vmm/vmm-reference/blob/main/src/api/Cargo.toml#L8

**Review by andreeaflorescu — APPROVED** [review]

File: src/vmm/src/lib.rs [hunk 1]
```diff
@@ -584,11 +602,11 @@ impl Vmm {
             DEFAULT_ADDRESSS_ALIGNEMNT,
             DEFAULT_ALLOC_POLICY,
         )?;
-
+        let irq = self.irq_allocator.next_irq()?;
         let mmio_range = mmio_from_range(&range);
         let mmio_cfg = MmioConfig {
             range: mmio_range,
-            gsi: 5,
+            gsi: irq,
```
Comment (lauralt) [inline src/vmm/src/lib.rs:609]: Can we check in some test (or add a new one) that the irq values are indeed the ones that were hardcoded before (e.g. check that after adding a block device, its irq is 5).

File: src/vmm/src/irq_allocator.rs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        assert_eq!(irq_alloc.next_irq(), Ok(5));
+
+        let _ = irq_alloc.next_irq();
+        assert_eq!(irq_alloc.next_irq(), Ok(7));
+
+        assert_eq!(irq_alloc.next_irq(), Err(Error::MaxIrq));
+
+        let mut irq_alloc = IrqAllocator::new(u32::MAX - 1, u32::MAX).unwrap();
+        assert_eq!(irq_alloc.next_irq(), Ok(u32::MAX));
+        assert!(irq_alloc.next_irq().is_err())
```
[the remark this answers is not in the document]
Comment (ramyak-mehra) [reply src/vmm/src/irq_allocator.rs:94]: this wasn't possible because line 29 prevents to add the values required to test overflow thats why it checks for an error rather than exact error, ideally overflow should never be reached because of how IrqAllocator is initialised

[the remark this answers is not in the document]
Comment (lauralt) [reply src/vmm/src/irq_allocator.rs:94]: Hmm, but this check only happens when initializing an irq allocator, you can then call next_irq(), and it is possible for last_used_irq and last_irq to have the same values (u32::MAX in this case).
If you replace this line with `assert_eq!(irq_alloc.next_irq(), Err(Error::IRQOverflowed));`, everything compiles and works just fine.

File: src/vmm/src/irq_allocator.rs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        Ok(IrqAllocator {
+            last_used_irq,
+            last_irq,
+        })
+    }
+
+    pub fn next_irq(&mut self) -> Result<u32> {
+        // The condition for overflow will never be reached because
+        // last_used_irq is always less than last_irq. So we can't have
+        // u32::MAX as last_used_irq and last_irq value, which would make it overlfow.
```
[the remark this answers is not in the document]
Comment (lauralt) [reply src/vmm/src/irq_allocator.rs:41]: The comment is not removed.

**Review by lauralt — APPROVED** [review]

**Review by andreeaflorescu — APPROVED** [review]
